src/Device/Driver/FreeVario.cpp

```diff
@@ -0,0 +1,297 @@
+#include "Device/Driver/FreeVario.hpp"
+#include "Device/Driver.hpp"
+#include "Device/Util/NMEAWriter.hpp"
+#include "Message.hpp"
+#include "NMEA/Checksum.hpp"
+#include "NMEA/Info.hpp"
+#include "NMEA/Derived.hpp"
+#include "NMEA/MoreData.hpp"
+#include "NMEA/InputLine.hpp"
+#include "Units/System.hpp"
+#include "Operation/Operation.hpp"
+#include "LogFile.hpp"
+#include "Interface.hpp"
+#include "CalculationThread.hpp"
+#include "Protection.hpp"
+
+/*
+ * Commands via NMEA from FreeVario Device to XCSoar:
+ *
+ * $PFV,M,S,<double value> - set MC External to defined value
+ * $PFV,M,U                - set MC External up by 0.1
+ * $PFV,M,D                - set MC External down by 0.1
+ *
+ * $PFV,F,C                - set Vario FlightMode to Circle
+ * $PFV,F,S                - set Vario FlightMode to SpeedToFly
+ *
+ * $PFV,Q,S,<double value> - set QNH to defined value
+ * $PFV,B,S,<double value> - set Bugs to percent by capacity 0 - 100
+ *
+ */
+
+class FreeVarioDevice : public AbstractDevice {
+    Port &port;
+
+public:
+  explicit FreeVarioDevice(Port &_port):port(_port) {}
+  bool ParseNMEA(const char *line,NMEAInfo &info) override;
+  static bool PFVParser(NMEAInputLine &line, NMEAInfo &info,Port &port);
+  bool PutMacCready(double mc, OperationEnvironment &env) override;
+  bool PutBugs(double bugs, OperationEnvironment &env) override;
+  bool PutQNH(const AtmosphericPressure &pres,OperationEnvironment &env) override;
+  void OnCalculatedUpdate(const MoreData &basic, const DerivedInfo &calculated) override;
+  void OnSensorUpdate(const MoreData &basic) override;
+};
+
+/**
+ * Parse NMEA messsage and check if it is a valid FreeVario message
+ * Is true when a valid message or false if no valid message
+ */
+bool
+FreeVarioDevice::PFVParser(NMEAInputLine &line, NMEAInfo &info, Port &port)
+{
+  NullOperationEnvironment env;
+  bool validMessage = false;
+
+  while (!line.IsEmpty()) {
+
+    char type = line.ReadOneChar();
+    char subCommand = line.ReadOneChar();
+
+    if (type == '\0')break;
+    if (subCommand == '\0')break;
+
+       switch (type) {
```

Comment:
The indentation gives the impression that the switch is in the body of the preceding if clause, which it isn't.
```suggestion
    if (subCommand == '\0')
      break;
    
    switch (type) {
```